src/Host.cpp

```diff
@@ -225,59 +230,478 @@ void Host::saveModules(int sync)
+                    // At some point between 0.11 and 1.x zip_source_close()
+                    // became public but it definetly is not available in 0.10
+                    // so fall back to older (void) function that has the same
+                    // effect but must only be used on an *unused* "zip_source"
+                    zip_source_free(zipSource);
+                    archive = 0;
+                    // Try again for next module...
+                    continue;
+                } else {
+                    // OK - file was found in archive - so now to replace it
+                    // Repurpose the index value to reflect the error status, 0 is
+                    // on success, -1 is error - the new libzip code also returns -1
+                    // on error and 0 or greater (being the index of the file in the
+                    // archive...
+                    index = zip_replace(archive, index, zipSource);
+                }
+
+                // index is -1 on error:
+                if (index == -1) {
+                    int ze;
+                    int se;
+                    char errorMessageBuffer[128];
+                    zip_error_get(archive, &ze, &se);
+                    zip_error_to_str(errorMessageBuffer, sizeof(errorMessageBuffer), ze, se);
+                    postMessage(tr("[ ERROR ] - Failed to replace file \"%1\" in module (archive)\n"
+                                               "file \"%2\",\n"
+                                               "error is: \"%3\".")
+                                .arg(filename_xml, moduleData.at(0), QString::fromUtf8(errorMessageBuffer)));
+                    zip_source_free(zipSource);
+                    zip_close(archive);
+                    archive = 0;
+                    // Try again for next module...
+                    continue;
+                }
+
+                if (zip_close(archive) == -1) {
+                    int ze;
+                    int se;
+                    char errorMessageBuffer[128];
+                    zip_error_get(archive, &ze, &se);
+                    zip_error_to_str(errorMessageBuffer, sizeof(errorMessageBuffer), ze, se);
+                    postMessage(tr("[ ERROR ] - Failed to replace file \"%1\" in module (archive)\n"
+                                               "file \"%2\",\n"
+                                               "error is: \"%3\".")
+                                .arg(filename_xml, moduleData.at(0), QString::fromUtf8(errorMessageBuffer)));
+                    // There is an unavoidable memory leak here for < 0.11 - we
+                    // cannot close the archive and there is no command to
+                    // forcibly clear the data associated with the archive data
+                    // structure.
+                } // End of error on closing archive (at the right time)
+            } // End of zipName not being empty
         }
-    }
+#endif
```

Comment:
Could we have the code for 3 different libzip versions be split out? This is one humongous function that does too many things!

Reply:
Humm, am trying that and it is workable - but making further changes will be harder to do I think to ensure they are applied uniformly - as much as possible - between the three...